### ISSUE TITLE
A running vm cannot dynamically scale cpu and memory.  It can be expanded when stopped   

### ISSUE THREAD

**Issue by 1450793561** [apache/cloudstack#6865]
#### Questrion:
> A running vm cannot dynamically scale cpu and memory.It's done when stopped

log detail

```
2022-11-02 14:44:13,311 INFO  [c.c.c.ClusterManagerImpl] (Cluster-Heartbeat-1:ctx-57dd945c) (logid:00501acf) Management node 1 is detected inactive by timestamp but is pingable
2022-11-02 14:44:13,629 DEBUG [c.c.a.ApiServlet] (qtp1462044018-848:ctx-5aeed521) (logid:1b967db1) ===START===  10.10.130.13 -- GET  virtualmachineid=beb97c61-bf41-4b38-ad74-1c65f6c5b125&keyword=&page=1&pageSize=10&details=min&response=json&command=listServiceOfferings
2022-11-02 14:44:13,649 DEBUG [c.c.a.ApiServer] (qtp1462044018-848:ctx-5aeed521 ctx-f8220033) (logid:1b967db1) CIDRs from which account 'Acct[2135826d-4dcd-11ed-89f1-02e685806f48-admin] -- Account {"id": 2, "name": "admin", "uuid": "2135826d-4dcd-11ed-89f1-02e685806f48"}' is allowed to perform API calls: 0.0.0.0/0,::/0
2022-11-02 14:44:13,663 WARN  [c.c.a.d.ParamGenericValidationWorker] (qtp1462044018-848:ctx-5aeed521 ctx-f8220033) (logid:1b967db1) Received unknown parameters for command listServiceOfferings. Unknown parameters : details
```
 
![微信图片_20221102150406](https://user-images.githubusercontent.com/31468018/199422235-1f0a256a-23d9-4877-a980-b5c7064d1660.png)

**Comment by DaanHoogland** [comment]
yes @1450793561 , this is a known limitation. I thought we had made sure the button was greyed or even omitted when the VM is running. What version did you encounter this with?

**Comment by 1450793561** [comment]
@DaanHoogland 
**CloudStack 4.17.1.0**

**Comment by DaanHoogland** [comment]
@1450793561 I don´t know if it is on someones schedule, so i'm afraid not.

**Comment by 1450793561** [comment]
yeh,ok

**Comment by GutoVeronezi** [comment]
Hello @1450793561 @weizhouapache

KVM requires defining ranges for CPU and memory in the VM XML to allow the VM to be dynamically scaled. Currently, those ranges are set only when working with custom compute offerings. For custom-constrained offerings, the ranges are defined while creating them; for custom-unconstrained offerings, the ranges are defined via the global settings `vm.serviceoffering.cpu.cores.max` and `vm.serviceoffering.ram.size.max`, or, if the settings are `0`, they are defined according to the host capacities.

As KVM supports dynamic scaling, this situation is a matter of designing the logical solution in ACS. We are planning to extend ACS to allow dynamic scaling VMs with fixed offerings in KVM; however, we do not have a due date yet.

---

@1450793561, while a VM with fixed offering cannot be dynamically scaled yet, you could try using custom offerings. Along with the custom offerings, some configurations have to be done:
- in the global settings, `enable.dynamic.scale.vm` must be `true`;
- the template and the VM must be marked as `Dynamically scalable`;

VMs started after setting those configurations will be able to be dynamically scaled. If you have any doubts or issues using this alternative, just ping me in this issue.

**Comment by weizhouapache** [comment]
> Hello @1450793561 @weizhouapache
> 
> KVM requires defining ranges for CPU and memory in the VM XML to allow the VM to be dynamically scaled. Currently, those ranges are set only when working with custom compute offerings. For custom-constrained offerings, the ranges are defined while creating them; for custom-unconstrained offerings, the ranges are defined via the global settings `vm.serviceoffering.cpu.cores.max` and `vm.serviceoffering.ram.size.max`, or, if the settings are `0`, they are defined according to the host capacities.
> 
> As KVM supports dynamic scaling, this situation is a matter of designing the logical solution in ACS. We are planning to extend ACS to allow dynamic scaling VMs with fixed offerings in KVM; however, we do not have a due date yet.
> 
> @1450793561, while a VM with fixed offering cannot be dynamically scaled yet, you could try using custom offerings. Along with the custom offerings, some configurations have to be done:
> 
> * in the global settings, `enable.dynamic.scale.vm` must be `true`;
> * the template and the VM must be marked as `Dynamically scalable`;
> 
> VMs started after setting those configurations will be able to be dynamically scaled. If you have any doubts or issues using this alternative, just ping me in this issue.

Great, thanks @GutoVeronezi for your detailed explanation.
Looking forward to the improvement.

**Comment by CodeBleu** [comment]
@rohityadavcloud This is a pretty important feature and was wondering if this could be placed in a milestone and assigned?

**Comment by GutoVeronezi** [comment]
Hello @CodeBleu,

We had some other priorities in the past 2 years and, unfortunately, we were not able develop into this improvement so far. However, we are planning to implement it in the next months, probably being addressed publicly on 4.21. We will update you as we make progress on this matter. 

We already have some use-cases to base the improvement on; however, feel free to share your opinion on how you expect this feature to work.

**Comment by DaanHoogland** [comment]
@bernardodemarco any updates on this? (as in will it make any of the upcoming releases 20.2 or 22?) ((happy to move it to 23 if nobody has time to implement))

**Comment by bernardodemarco** [comment]
Hello, @DaanHoogland 

> @bernardodemarco any updates on this?

Yes, I've started implementing it in the past few days. I'm not sure it'll be ready for the `4.22.0` release, but I'll try

**Comment by weizhouapache** [comment]
@bernardodemarco 
any update on this ?

**Comment by bernardodemarco** [comment]
Hello, @weizhouapache 

Sorry for the delay, I was on vacation. I'll try to open the PR, along with the feature's specification, until the end of this month.

**Comment by bernardodemarco** [comment]
Hello, all

I have just published the feature specification in #12908. Please feel free to review it and share your thoughts.

I will aim to open the PR at the beginning of next week.

Best regards,
Bernardo De Marco Gonçalves
